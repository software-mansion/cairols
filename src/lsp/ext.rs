//! CairoLS extensions to the Language Server Protocol.

use std::path::PathBuf;

use lsp_types::notification::Notification;
use lsp_types::request::Request;
use lsp_types::{TextDocumentPositionParams, Url};
use serde::{Deserialize, Serialize};

/// Provides content of virtual file from the database.
pub(crate) struct ProvideVirtualFile;

#[derive(Debug, Eq, PartialEq, Clone, Deserialize, Serialize)]
pub(crate) struct ProvideVirtualFileRequest {
    pub uri: Url,
}

#[derive(Debug, Eq, PartialEq, Clone, Deserialize, Serialize)]
pub(crate) struct ProvideVirtualFileResponse {
    pub content: Option<String>,
}

impl Request for ProvideVirtualFile {
    type Params = ProvideVirtualFileRequest;
    type Result = ProvideVirtualFileResponse;
    const METHOD: &'static str = "vfs/provide";
}

/// Collects information about all Cairo crates that are currently being analyzed.
pub struct ViewAnalyzedCrates;

impl Request for ViewAnalyzedCrates {
    type Params = ();
    type Result = String;
    const METHOD: &'static str = "cairo/viewAnalyzedCrates";
}

/// Provides string with code after macros expansion.
pub struct ExpandMacro;

impl Request for ExpandMacro {
    type Params = TextDocumentPositionParams;
    type Result = Option<String>;
    const METHOD: &'static str = "cairo/expandMacro";
}

/// Notifies about corelib version mismatch.
#[derive(Debug)]
pub struct CorelibVersionMismatch;

impl Notification for CorelibVersionMismatch {
    type Params = String;
    const METHOD: &'static str = "cairo/corelib-version-mismatch";
}

/// Collects versions of LS and it's dependencies.
#[derive(Debug)]
pub struct ToolchainInfo;

#[derive(Serialize, Deserialize)]
pub struct PathAndVersion {
    pub path: PathBuf,
    pub version: String,
}

#[derive(Serialize, Deserialize)]
pub struct ToolchainInfoResponse {
    pub ls: PathAndVersion,
    pub scarb: Option<PathAndVersion>,
}

impl Request for ToolchainInfo {
    type Params = ();
    type Result = ToolchainInfoResponse;
    const METHOD: &'static str = "cairo/toolchainInfo";
}

#[cfg(feature = "testing")]
pub mod testing {
    use lsp_types::notification::Notification;

    /// Notifies about the end of project updating.
    #[derive(Debug)]
    pub struct ProjectUpdatingFinished;

    impl Notification for ProjectUpdatingFinished {
        type Params = ();
        const METHOD: &'static str = "cairo/projectUpdatingFinished";
    }

    /// Notifies about diagnostics generation which is beginning to calculate.
    #[derive(Debug)]
    pub struct DiagnosticsCalculationStart;

    impl Notification for DiagnosticsCalculationStart {
        type Params = ();
        const METHOD: &'static str = "cairo/diagnosticsCalculationStart";
    }

    /// Notifies about diagnostics generation which ended calculating.
    #[derive(Debug)]
    pub struct DiagnosticsCalculationFinish;

    impl Notification for DiagnosticsCalculationFinish {
        type Params = ();
        const METHOD: &'static str = "cairo/diagnosticsCalculationFinish";
    }
}

#[derive(Debug)]
pub struct ScarbPathMissing {}

impl Notification for ScarbPathMissing {
    type Params = ();
    const METHOD: &'static str = "scarb/could-not-find-scarb-executable";
}

#[derive(Debug)]
pub struct ScarbResolvingStart {}

impl Notification for ScarbResolvingStart {
    type Params = ();
    const METHOD: &'static str = "scarb/resolving-start";
}

#[derive(Debug)]
pub struct ScarbResolvingFinish {}

impl Notification for ScarbResolvingFinish {
    type Params = ();
    const METHOD: &'static str = "scarb/resolving-finish";
}
