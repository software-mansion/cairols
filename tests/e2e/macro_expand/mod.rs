use cairo_lang_utils::ordered_hash_map::OrderedHashMap;
use cairo_language_server::lsp::ext::ExpandMacro;
use lsp_types::{TextDocumentIdentifier, TextDocumentPositionParams};
use serde::Serialize;

use crate::support::cairo_project_toml::CAIRO_PROJECT_TOML_2024_07;
use crate::support::cursor::peek_caret;
use crate::support::data::data;
use crate::support::{cursors, sandbox};

#[test]
fn test_macro_expand_attribute() {
    check(data!("attribute.cairo.template"), "attribute")
}

#[test]
fn test_macro_expand_simple_inline() {
    check(data!("simple_inline.cairo.template"), "simple_inline")
}

#[test]
fn test_macro_expand_empty() {
    check(data!("empty.cairo.template"), "empty")
}

#[test]
fn test_macro_expand_derive() {
    check(data!("derive.cairo.template"), "derive")
}

fn check(code: String, case_name: &str) {
    let (cairo, cursors) = cursors(&code);

    let mut ls = sandbox! {
        files {
            "cairo_project.toml" => CAIRO_PROJECT_TOML_2024_07,
            "src/lib.cairo" => cairo.clone(),
        }
    };

    ls.open_all_cairo_files_and_wait_for_project_update();

    let mut ordered_hash_map: OrderedHashMap<String, Vec<String>> = OrderedHashMap::default();
    cursors.carets().into_iter().for_each(|position| {
        let source_context = peek_caret(&cairo, position);

        let macro_expansion = ls.send_request::<ExpandMacro>(TextDocumentPositionParams {
            position,
            text_document: TextDocumentIdentifier { uri: ls.doc_id("src/lib.cairo").uri },
        });

        let expansion =
            macro_expansion.unwrap_or_else(|| String::from("No expansion information.\n"));

        ordered_hash_map.entry(expansion).or_default().push(source_context);
    });

    for (snap_seq, (expansion, source_contexts)) in ordered_hash_map.into_iter().enumerate() {
        let source_contexts = source_contexts.join("\n");
        let report = MacroExpandReport { source_contexts, expansion };

        insta::assert_toml_snapshot!(format!("{case_name}#snap_{snap_seq}"), report);
    }
}

#[derive(Serialize)]
struct MacroExpandReport {
    source_contexts: String,
    expansion: String,
}
