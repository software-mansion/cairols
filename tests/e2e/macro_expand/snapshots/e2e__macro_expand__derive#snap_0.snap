---
source: tests/e2e/macro_expand/mod.rs
expression: report
---
source_contexts = '''
#[d<caret>erive(Drop, Serde)]

#[derive(Drop,<caret> Serde)]

#[derive(Drop, Serde)]<caret>

struct <caret>A {
'''
expansion = '''
// lib.cairo
// ---------

#[derive(Drop, Serde)]
struct A {
    a: felt252,
}

// impls
// -----

impl ADrop<> of core::traits::Drop<A>;
impl ASerde<> of core::serde::Serde<A> {
    fn serialize(self: @A, ref output: core::array::Array<felt252>) {
        core::serde::Serde::<felt252>::serialize(self.a, ref output)
    }
    fn deserialize(ref serialized: core::array::Span<felt252>) -> core::option::Option<A> {
        let a = core::internal::InferDestruct::<
            felt252,
        > { value: core::serde::Serde::<felt252>::deserialize(ref serialized)? };
        core::option::Option::Some(A { a: a.value })
    }
}'''
